### DISCUSSION TITLE
How to output stdout into a box?

### DISCUSSION THREAD

**Discussion by mscofield0** [ArthurSonzogni/FTXUI#273] [Q&A] [answered]
I would like to create a chat app, where I have the main message window, an input window and an additional "info" window that just displays in whatever gets sent to stdout and stderr normally.

**Comment by ArthurSonzogni** [comment, marked as answer]
FTXUI reads from stdin, and write to stdout. So, if you write to stdout directly it would break the interface.

So I recommand not writting to stdout directly, but create your special set of function to write elsewhere.

(Note: One can also try to reroute file descriptor. For instance I redirected [stdin](https://github.com/ArthurSonzogni/chrome-log-beautifier/blob/master/src/main.cpp#L44) here)

Let's say the data is stored into:
```cpp
std::string data;
```

Then, you can update the data:
```cpp
void Write(const std::string& new_data) {
  data += new_data;
}
```

Then you need an DOM Element drawing it:
```
auto element = text(data);
```

I guess, you want a component producing the DOM element for every new frame. You can use:
```cpp
auto component = Renderer([&data] {
  return text(data)
});
```

A `text` only support drawing a single line. You probably don't like this. So you can instead store std::vector<std::string> data and draw:
```cpp
Elements elements;
for(auto& line : data) {
  elements.push_back(text(line));
}
return vbox(std::move(elements));
```

Or you can continue to store in a std::string and break by new lines dynamically:
```cpp
auto component = Renderer([&data] {
  // Break `data` by new lines:
  auto lines = lines();
  auto start = 0U;
  auto end = s.find('\n');
  while (end != std::string::npos) {
    lines.push_back(text(s.substr(start, end-start)));
    start = end + 1;
    end = s.find('\n', start);
  }
  lines.push_back(text(s.substr(start, end)));

  return vbox(std::move(lines));
});

```

**Reply by nullbyto** [reply]
Thanks it helped a lot!
I'm trying to make a box with text getting updated which works, but also i need to make it scrollable which i can't figure out how to do it correctly, it just makes draws a scrollbar but it cant be controlled. Is there any way to make such scrollable text area ?

**Comment by ArthurSonzogni** [comment]
Maybe you can take Scroller from:
https://github.com/ArthurSonzogni/git-tui/tree/master/src
and wrap your component?

**Reply by nullbyto** [reply]
i will try that thank you :D
